lib/bandit/http1/adapter.ex

```diff
@@ -75,9 +93,9 @@ defmodule Bandit.HTTP1.Adapter do
 
     case :erlang.decode_packet(type, req.buffer, packet_size: packet_size) do
       {:more, _len} ->
-        with {:ok, iodata} <- read(req.socket, 0) do
+        with {:ok, chunk} <- read_available(req.socket, _read_timeout = nil) do
           # decode_packet expects a binary, so convert it to one
-          req = %{req | buffer: IO.iodata_to_binary([req.buffer | iodata])}
+          req = %{req | buffer: req.buffer <> chunk}
```

Comment:
Why the concat?

Reply:
Because `chunk` is guaranteed to be a binary here.

Reply:
Okay so assume (correctly) that I don't know why that makes it better.

As my understanding goes: string concat is O(n) and adding to improper list is O(1).
I know big-O by itself isn't always the best way to compare performance and there are subtleties if you know ahead of time what your data looks like, so I was hoping you could explain why concat would be better here. Is it just because you use `byte_size/1` later which is better than `IO.iodata_length/1`? And if so why the swap? Or is there more to it?

Reply:
You mean that snippet?

```elixir
        result = IO.iodata_to_binary([buffer | iolist])
        result_size = byte_size(result)
        body_remaining = body_remaining - result_size - byte_size(buffer)
```

`byte_size/1` executes in constant time irrelevant to the size of the binary, while `IO.iodata_length/1` has to traverse entire the iodata list for the second time (the first time was when `IO.iodata_to_binary/1` converted the iodata to a binary and stored its size).

```elixir
          req = %{req | buffer: IO.iodata_to_binary([req.buffer | iodata])}
          req = %{req | buffer: req.buffer <> chunk}
```

Yes, indeed - appending to improper list is O(1) + an extra word allocation and if that data would go straight over the wire, that would be all complexity. But `req.buffer` is a binary and freshly constructed list need to be converted to a binary with `IO.iodata_to_binary/1`. What is complexity of `IO.iodata_to_binary/1`? 

On the other size `<>/2` compiles to bit syntax like `<<Buffer/binary, Chunk/binary>>` and according to the documentation (https://www.erlang.org/doc/efficiency_guide/binaryhandling.html#constructing-binaries) BEAM might enable an optimization which would prevent those binaries from being copied by placing `Chunk` in to `Buffer`.

Reply:
Thanks and yeah this line of code just clicked to me why it is obviously better.

Reply:
https://github.com/erlang/otp/blob/maint-26/erts/emulator/beam/binary.c#L928 Erlang's implementation of iolist_to_binary BIF. Sorry, I don't have much time at the moment to study it.

Reply:
I need to stop reading PRs in the morning before coffee.